src/tip-opengov.ts

```diff
@@ -29,40 +31,73 @@ export async function tipOpenGov(opts: {
   if ("error" in track) {
     return { success: false, errorMessage: track.error };
   }
+  const contributorAddress = contributor.account.address;
 
   const proposalTx = api.tx.utility.batch([
     api.tx.system.remark(formatReason(tipRequest)),
-    api.tx.treasury.spend(track.value.toString(), contributor.account.address),
+    api.tx.treasury.spend(track.value.toString(), contributorAddress),
   ]);
   const encodedProposal = proposalTx.method.toHex();
   const proposalHash = blake2AsHex(encodedProposal);
+  const encodedLength = Math.ceil((encodedProposal.length - 2) / 2);
 
-  const preimage_unsub = await api.tx.preimage
-    .notePreimage(encodedProposal)
-    .signAndSend(botTipAccount, { nonce: -1 }, (result) => {
-      if (result.status.isInBlock) {
-        bot.log(`Preimage Upload included at blockHash ${result.status.asInBlock.toString()}`);
-      } else if (result.status.isFinalized) {
-        bot.log(`Preimage Upload finalized at blockHash ${result.status.asFinalized.toString()}`);
-        preimage_unsub();
-      }
-    });
+  return await new Promise(async (resolve, reject) => {
+    // create a preimage from opengov with the encodedProposal above
+    const preimageUnsubscribe = await api.tx.preimage
+      .notePreimage(encodedProposal)
+      .signAndSend(botTipAccount, { nonce: -1 }, async (result) => {
+        await signAndSendCallback(bot, contributor.account, "preimage", preimageUnsubscribe, result)
+          .then(async () => {
+            const readPreimage = await api.query.preimage.statusFor(proposalHash);
 
-  const referenda_unsub = await api.tx.referenda
-    .submit(
-      // TODO: There should be a way to set those types properly.
-      { Origins: track.track } as any, // eslint-disable-line
-      { Lookup: { hash: proposalHash, length: proposalTx.length - 1 } },
-      { after: 10 } as any, // eslint-disable-line
-    )
-    .signAndSend(botTipAccount, { nonce: -1 }, (result) => {
-      if (result.status.isInBlock) {
-        bot.log(`Tip referendum included at blockHash ${result.status.asInBlock.toString()}`);
-      } else if (result.status.isFinalized) {
-        bot.log(`Tip referendum finalized at blockHash ${result.status.asFinalized.toString()}`);
-        referenda_unsub();
-      }
-    });
+            if (readPreimage.isEmpty) {
+              reject(new Error(`Preimage for ${proposalHash} was not found, check if the bot has enough funds.`));
+            }
 
-  return { success: true, tipUrl: getTipUrl(contributor.account.network) };
+            const proposalUnsubscribe = await api.tx.referenda
+              .submit(
+                // TODO: There should be a way to set those types properly.
+                { Origins: track.track } as never,
+                { Lookup: { hash: proposalHash, len: encodedLength } },
```

Comment:
```suggestion
                { Lookup: { hash: proposalHash, len: readPreimage.len } },
```
Not sure if that is the correct syntax, but the len should be in the `RequestStatus`

![image](https://github.com/paritytech/substrate-tip-bot/assets/10380170/d42108ea-4177-47c0-86a5-3c1c8af31d1d)

You can also SCALE encode yourself or use the encoded len though.

Reply:
`len: readPreimage.len` there's no such field in the returned response, there's `readPreimage.encodedLength` but it returns 54, while expected number is 85

Here's some logs

```
proposalTx.encodedLength 88
proposalTx.toString {"signature":{"signer":{"id":"CaKWz5omakTK7ovp4m3koXrHyHb7NG3Nt7GENHbviByZpKp"},"signature":{"ed25519":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"},"era":{"immortalEra":"0x00"},"nonce":0,"tip":0},"method":{"callIndex":"0x1800","args":{"calls":[{"callIndex":"0x0000","args":{"remark":"0x544f3a206d6f7264616d617820464f523a206d6567617265706f2331342028736d616c6c29"}},{"callIndex":"0x1203","args":{"amount":2000000000000,"beneficiary":{"id":"HnMAUz7r2G8G3hB27SYNyit5aJmh2a5P4eMdDtACtMFDbam"}}}]}}}
proposalTx.length 86
encodedProposal 0x180008000094544f3a206d6f7264616d617820464f523a206d6567617265706f2331342028736d616c6c2912030b00204aa9d10100e659a7a1628cdd93febc04a4e0646ea20e9f5f0ce097d9a05290d4a9e054df4e
encodedProposal.length 172
readPreimage.encodedLength 54 (¯\\_(ツ)_/¯)
encodedLength 85
```

when I create it, in polkadotjs it shows 85

![image](https://github.com/paritytech/substrate-tip-bot/assets/1177472/f108accd-06d7-4d32-baf6-9779d20e8879)

Reply:
I just added 5 chars to a remark, and got 90 as final size in polkadotjs

so it means to me 
```
proposalTx.encodedLength - 3
proposalTx.length - 1
Math.ceil((encodedProposal.length - 2) / 2)
```

all 3 will give same correct size of preimage

https://github.com/search?q=+%22Math.ceil%28%28encodedProposal.length+-+2%29+%2F+2%29%3B%22&type=code

I just used last as this is most commonly used

Reply:
Okay yea then just use what works. The TS code is fucked either way.